pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -813,9 +815,17 @@ protected ByteBuf encryptMessage(MessageMetadata msgMetadata, ByteBuf compressed
         }
     }
 
-    protected ByteBufPair sendMessage(long producerId, long sequenceId, int numMessages, MessageMetadata msgMetadata,
-            ByteBuf compressedPayload) {
-        return Commands.newSend(producerId, sequenceId, numMessages, getChecksumType(), msgMetadata, compressedPayload);
+    protected ByteBufPair sendMessage(long producerId, long sequenceId, int numMessages,
+                                      MessageId messageId, MessageMetadata msgMetadata,
+                                      ByteBuf compressedPayload) {
+        if (messageId instanceof MessageIdImpl) {
```

Comment:
This shouldn't be a code path taken by user Pulsar clients but only by the Replicator.

Maybe we should add an internal flag to the PulsarClient to mark it as a internal client and allow it to be able to do this thing.
Pushing the messageId from the producer also may lead to some security issues.
We must prevent on the broker side that a client writes to the wrong ledger (for instance to a ledger of another topic, that  can be not accessible from the client according the authz)

Reply:
So we have two problems:
- regular users should not be able to set the MessageId (I don't know if there is a way)
- the broker must not allow a client to write to a ledger that is not part of the topic

Reply:
@eolivelli 

> * regular users should not be able to set the MessageId (I don't know if there is a way)

For normal users, the public method `TypedMessageBuilder<T> Producer#newMessage()` should be used to construct and send messages, but they can't set MessageId in `TypedMessageBuilder`. So messageId won't be set unless users know what they are doing.
And currently, messageId only can be send to broker with public methods in `ProducerImpl`. It's already kind of an internal usage.

> * the broker must not allow a client to write to a ledger that is not part of the topic

Sure, I understand your concern, but it won't happen. The messageId will only be used by ShadowTopic and they won't have any write access to these ledgers or any ledgers. For normal topic, in case they received this messageId by some accident, they will just ignore it.

Reply:
@eolivelli PTAL

Reply:
It's very confusing to see the logic here. What's the behavior if the ledger id and the entry id are both -1 in a `CommandSend`? Is this `else` branch protected by unit tests? @Jason918 @eolivelli